### PR TITLE
[WIP] Implement spans exporting for ClickHouse storage for Jaeger V2

### PR DESCRIPTION
**Pull request by haanhvu**
## Which problem is this PR solving?
First implementation step of #4196 for Jaeger V2

## How was this change tested?
Still in WIP. Add tests later.

### CODE REVIEW

**Comment by haanhvu** [comment]
@yurishkuro This still needs some supplements and a lot of cleanups and tests. But could you take an initial look to see if I'm going the right way? Thanks!

File: cmd/jaeger/internal/extension/jaegerstorage/config.go [hunk 1]
```diff
@@ -13,9 +14,16 @@ type Config struct {
...
+
+	ClickHouse map[string]ch.Config `mapstructure:"clickhouse"`
 }
 
 type MemoryStorage struct {
 	Name string `mapstructure:"name"`
 	memoryCfg.Configuration
 }
+
+type ClickHouseStorage struct {
```
Comment (yurishkuro) [inline cmd/jaeger/internal/extension/jaegerstorage/config.go:26]: is this type used?

File: plugin/storage/clickhouse/factory.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	} else {
+		f.client = client
+	}
+
+	return f
+
+	//TODO: Move some steps to Initialize()
+}
+
+func (f *Factory) CreateSpansTable(ctx context.Context) {
```
Comment (yurishkuro) [inline plugin/storage/clickhouse/factory.go:43]: why is this function public, who would call it?

File: plugin/storage/clickhouse/factory.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	if err := chSpanStore.CreateSpansTable(ctx, f.client, f.spansTableName); err != nil {
+		f.logger.Error("failed to create spans table", zap.Error(err))
+	}
+}
+
+func (f *Factory) ExportSpans(ctx context.Context, td ptrace.Traces) error {
```
Comment (yurishkuro) [inline plugin/storage/clickhouse/factory.go:49]: why is ExportSpans defined on the factory instead of span store?

Comment (haanhvu) [reply plugin/storage/clickhouse/factory.go:49]: The logic is defined in `plugin/storage/spanstore/clickhouse/exporter.go` ([link](https://github.com/jaegertracing/jaeger/pull/4941/files#diff-e85bace6bfd6e03dd602c1d1bd2ec27367e539b554e541d6f30786a63d652f2eR43)). But the exporter is depending on the real type of Factory to decide whether to call `SpanWriter.WriteSpan()` or `ExportSpans()` ([link](https://github.com/jaegertracing/jaeger/pull/4941/files#diff-d11f17877ed77e35143c698010924340c6cf95489656e3ccb0b3c2654c9b9853R42-R52). So I think making `ExportSpans()` a func of clickhouse Factory may be a good idea.

Of course, a better idea is to define a new API for `ExportSpans()`, but maybe we can do that later? Or you have a better idea?

File: cmd/jaeger/internal/extension/jaegerstorage/config.go (hunk 1 above)
Comment (haanhvu) [reply cmd/jaeger/internal/extension/jaegerstorage/config.go:26]: Ah, I followed your MemoryStorage struct above. Now realize I'm currently not needing this for anything. Do you use MemoryStorage struct for anything?

File: plugin/storage/clickhouse/factory.go (hunk 2 above)
Comment (haanhvu) [reply plugin/storage/clickhouse/factory.go:43]: Currently, Jaeger V2's exporter checks the type of Factory. If it's a clickhouse Factory, then f.CreateSpansTable() is called, as you can see in this [piece](https://github.com/jaegertracing/jaeger/pull/4941/files#diff-d11f17877ed77e35143c698010924340c6cf95489656e3ccb0b3c2654c9b9853R42-R52).

File: plugin/storage/clickhouse/factory.go (hunk 3 above)
Comment (yurishkuro) [reply plugin/storage/clickhouse/factory.go:49]: I have a better idea, but it is not well formed yet, I need to write it down and maybe put together the new interfaces. Will try to do this later this weekend. Are you blocked on it?

Comment (haanhvu) [reply plugin/storage/clickhouse/factory.go:49]: No I don't think this blocks the progress. Can continue.

**Comment by k0zl** [comment]
👍 I've been waiting clickhouse support for a long time. Is there anything I can help you with?

**Comment by haanhvu** [comment]
> 👍 I've been waiting clickhouse support for a long time. Is there anything I can help you with?

@k0zl Sure! For this PR (spans exporting), the remaining parts are:

- Finish schema implementation in `plugin/storage/clickhouse/spanstore/schema.go`. We described our schema choices in our [benchmark report](https://medium.com/jaegertracing/making-design-decisions-for-clickhouse-as-a-core-storage-backend-in-jaeger-62bf90a979d)). As you can see in the current `schema.go`, I just implemented the spans table, not the materialized views yet.
- Add tests: I'm pushing an unit `exporter_test.go` this week. There are other unit tests, e2e test, or even local test with Jaeger v2 binary to do. (Jaeger v2 issue/decription is [here](https://github.com/jaegertracing/jaeger/issues/4843))

After finishing with spans exporting, we'll move to spans reading,...

Also, since clickhouse storage is supported in Jaeger v2, to make this realeased you could also help with Jaeger v2 roadmap, in the issue I linked above.

Please choose whatever interests you. Thanks!

**Comment by yurishkuro** [comment]
@haanhvu it would be useful to write a design doc / roadmap documenting what we're doing and what tasks need to be done, are you up for that?  You can model it after the Jaeger V2 document.

**Comment by haanhvu** [comment]
> @haanhvu it would be useful to write a design doc / roadmap documenting what we're doing and what tasks need to be done, are you up for that? You can model it after the Jaeger V2 document.

Yeah I'll create an issue for supporting clickhouse in jaeger v2 tomorrow

**Comment by haanhvu** [comment]
@yurishkuro Could you take a look at the failed [all-in-one build](https://github.com/jaegertracing/jaeger/actions/runs/7358386059/job/20031566276?pr=4941)? `make all-in-one-integration-test` passed locally for me. Not sure why it fails in CI.

**Comment by yurishkuro** [comment]
I am seeing error
```
Error: failed to build pipelines: failed to create "jaeger_storage_exporter" exporter for data type "traces": nil PushTraces
2023/12/29 16:57:59 failed to build pipelines: failed to create "jaeger_storage_exporter" exporter for data type "traces": nil PushTraces
+ exit 1
```

File: cmd/jaeger/internal/extension/jaegerstorage/config.go (hunk 1 above)
Comment (yurishkuro) [reply cmd/jaeger/internal/extension/jaegerstorage/config.go:26]: not needed, you can remove both

File: cmd/jaeger/internal/exporters/storageexporter/exporter.go [hunk 4]
```diff
@@ -14,13 +14,16 @@ import (
...
 )
 
 type storageExporter struct {
-	config     *Config
-	logger     *zap.Logger
-	spanWriter spanstore.Writer
+	config             *Config
+	logger             *zap.Logger
+	spanWriter         spanstore.Writer
+	exportTraces       func(ctx context.Context, td ptrace.Traces) error
```
Comment (yurishkuro) [inline cmd/jaeger/internal/exporters/storageexporter/exporter.go:25]: nit: I would call this `consumeTraces`, because the argument that we pass to the exporter helper is `consumer.ConsumeTracesFunc`

File: cmd/jaeger/internal/exporters/storageexporter/factory.go [hunk 5]
```diff
@@ -34,8 +34,21 @@ func createDefaultConfig() component.Config {
 func createTracesExporter(ctx context.Context, set exporter.CreateSettings, config component.Config) (exporter.Traces, error) {
 	cfg := config.(*Config)
 	ex := newExporter(cfg, set.TelemetrySettings)
+
+	if ex.requireBatchInsert {
```
Comment (yurishkuro) [inline cmd/jaeger/internal/exporters/storageexporter/factory.go:38]: you don't need an if statement:
```go
exporterhelper.WithQueue(exporterhelper.QueueSettings{Enabled: ex.requireBatchInsert}),
```

File: cmd/jaeger/internal/exporters/storageexporter/exporter.go [hunk 6]
```diff
@@ -30,14 +33,23 @@ func newExporter(config *Config, otel component.TelemetrySettings) *storageExpor
...
+	case *ch.Factory:
+		t.CreateSpansTable(ctx)
+		exp.requireBatchInsert = true
+		exp.exportTraces = t.ExportSpans
+	default:
+		if exp.spanWriter, err = f.CreateSpanWriter(); err != nil {
+			return fmt.Errorf("cannot create span writer: %w", err)
+		}
+		exp.requireBatchInsert = false
+		exp.exportTraces = exp.pushTraces
```
Comment (yurishkuro) [inline cmd/jaeger/internal/exporters/storageexporter/exporter.go:52]: nit: I'd rename `s/pushTraces/writeSpans/` to indicate jaeger-v1 style storage

File: cmd/jaeger/internal/exporters/storageexporter/exporter.go [hunk 7]
```diff
@@ -30,14 +33,23 @@ func newExporter(config *Config, otel component.TelemetrySettings) *storageExpor
...
+func (exp *storageExporter) start(ctx context.Context, host component.Host) error {
 	f, err := jaegerstorage.GetStorageFactory(exp.config.TraceStorage, host)
 	if err != nil {
 		return fmt.Errorf("cannot find storage factory: %w", err)
 	}
 
-	if exp.spanWriter, err = f.CreateSpanWriter(); err != nil {
-		return fmt.Errorf("cannot create span writer: %w", err)
+	switch t := f.(type) {
+	case *ch.Factory:
```
Comment (yurishkuro) [inline cmd/jaeger/internal/exporters/storageexporter/exporter.go:43]: it seems you're doing this too late, I think `createTracesExporter` would be called before `start`

**Comment by haanhvu** [comment]
> I am seeing error
> 
> ```
> Error: failed to build pipelines: failed to create "jaeger_storage_exporter" exporter for data type "traces": nil PushTraces
> 2023/12/29 16:57:59 failed to build pipelines: failed to create "jaeger_storage_exporter" exporter for data type "traces": nil PushTraces
> + exit 1
> ```

@yurishkuro This error didn't come up in my `make all-in-one-integration-test`. Probably because my `make all-in-one-integration-test` checked v1, not v2? If that's the case how to specify v2 in this test locally?

**Comment by yurishkuro** [comment]
> This error didn't come up in my make `all-in-one-integration-test`

This test doesn't just run on its own, it expects the binary running, so how are you starting the binary?

For example, this works:
```shell
# terminal 1
$ go run -tags=ui ./cmd/jaeger

# terminal 2
$ SKIP_SAMPLING=true make all-in-one-integration-test
```

But this is running with default aio config for OTEL, whereas you need to provide your own config in order to test the same against CH.

**Comment by haanhvu** [comment]
> ```shell
> go run -tags=ui ./cmd/jaeger
> ```

Ah, I ran the binary with `make run-all-in-one`... That's why v2 wasn't checked... Let me try again

**Comment by haanhvu** [comment]
> But this is running with default aio config for OTEL, whereas you need to provide your own config in order to test the same against CH.

So I'll have to run clickhouse server, make my own jaeger v2 config (that exports to clickhouse) and `go run -tags=ui -config=config.yaml ./cmd/jaeger` right? I'll try it tomorrow.

**Comment by haanhvu** [comment]
@yurishkuro I ran the binary with my [config.yaml](https://github.com/jaegertracing/jaeger/pull/4941/files#diff-d8d0422389f03d783e32e627250fe29834bd09c6361640d1ff00661dd6820034), spans were exported successfully to clickhouse:
Screenshot from 2024-01-03 03-07-22

Should we add an e2e test now, or wait until spans reading are implemented?

Also, can you check the unit tests? I just have one `factory_test.go`, no test files for `config.go` or `spanstore` because factory already calls config and spanstore.

Is there anything else you think we should test in this PR?

**Comment by yurishkuro** [comment]
> Should we add an e2e test now, or wait until spans reading are implemented?

You wouldn't be able to do e2e without reading, that's how the tests verify that traces are saved.

File: cmd/jaeger/internal/exporters/storageexporter/exporter.go [hunk 8]
```diff
@@ -30,14 +35,22 @@ func newExporter(config *Config, otel component.TelemetrySettings) *storageExpor
...
+		exp.clickhouse = true
+		t.CreateSpansTable(ctx)
```
Comment (yurishkuro) [inline cmd/jaeger/internal/exporters/storageexporter/exporter.go:47]: why can't we do this inside `clickhouse.NewFactory`? `storageexporter` doesn't need to know about internal details of each storage implementation

**Comment by haanhvu** [comment]
@yurishkuro I rebased and made a little change following what we're having with other V2 storage. I checked the change locally and it worked:
Screenshot from 2024-06-13 15-15-16

There's one goleak failure in unit tests. However, the goleak failure is in `database/sql` package, not our code. Should we ignore or do what with this?

Also, the code coverage is a bit skewed. Seems like codecov checks per file or per package and crossed-files and cross-packages function calls are not counted. Do you want to rearrange the files/packages so that code coverage is more accurate?

Please also checks if we need more tests in this PR.

**Comment by yurishkuro** [comment]
> There's one goleak failure in unit tests. However, the goleak failure is in database/sql package, not our code. Should we ignore or do what with this?

We shouldn't ignore leaks. Most likely it's from not closing resources in tests properly, you can see which goroutine is remaining running and understand where it starts. There are rare cases where libraries introduce unsolvable leaks (like glog), but I am skeptical database/sql would be such lib.

The codecov for storage is usually combined from unit tests and e2e tests.

Note that V2 Storage API [already exists](https://github.com/jaegertracing/jaeger/pull/5497), and we have an adapter to v1 that will be used by the `jaegerexporter`. The Clickhouse implementation can be the first one to be natively supporting V2.

**Comment by haanhvu** [comment]
> Note that V2 Storage API https://github.com/jaegertracing/jaeger/pull/5497, and we have an adapter to v1 that will be used by the jaegerexporter. The Clickhouse implementation can be the first one to be natively supporting V2.

Yeah I saw the conflicts in `exporter.go`. I'll rebase and make the current implementation work with the rebase first, then make clickhouse v2-native next.

**Comment by haanhvu** [comment]
The latest rebased implementation works (not v2-native yet)
Screenshot from 2024-06-23 13-19-14

**Comment by JellyTony** [comment]
Is there any progress on this PR currently, and when will it be merged
